### PR TITLE
Auth protection improvements

### PR DESCRIPTION
**Pull request by lukasbestle**

### CODE REVIEW

File: config/api/routes/auth.php [hunk 1]
```diff
@@ -36,24 +36,21 @@
...
+                    'code'   => 200,
+                    'status' => 'ok',
+                    'user'   => $this->resolve($user)->view('auth')->toArray()
+                ];
             } catch (Throwable $e) {
                 if ($this->kirby()->option('debug') === true) {
                     throw $e;
                 } else {
-                    // catch any kind of login error
+                    // hide the actual error to avoid leaking security-relevant information
```
Comment (bastianallgeier) [inline config/api/routes/auth.php]: You basically undid my changes from the last version here. I added the option to get proper info about failed logins when Kirby is in debug mode. There were quite a few people who had issues with failed logins and couldn't debug them properly. I know about the security risk, but I still think we need a way to have proper error handling with useful error messages when you actually want to debug this.

File: i18n/translations/en.json [hunk 2]
```diff
@@ -169,7 +169,6 @@
   "error.user.password.notSame": "The passwords do not match",
   "error.user.password.undefined": "The user does not have a password",
   "error.user.role.invalid": "Please enter a valid role",
-  "error.user.undefined": "The user cannot be found",
```
Comment (bastianallgeier) [inline i18n/translations/en.json:172]: I wouldn't remove this. This is also a useful message for the panel and I think we even use it somewhere else.

File: src/Cms/Auth.php [hunk 3]
```diff
@@ -79,13 +80,11 @@ public function currentUserFromBasicAuth(BasicAuth $auth = null)
             throw new PermissionException('Basic authentication is only allowed over HTTPS');
         }
 
-        if ($user = $this->kirby->users()->find($auth->username())) {
-            if ($user->validatePassword($auth->password()) === true) {
-                return $user;
-            }
+        try {
+            return $this->validatePassword($auth->username(), $auth->password());
+        } catch (Throwable $e) {
```
Comment (bastianallgeier) [inline src/Cms/Auth.php]: Same as above. If we want to have a way to get proper error messages on failed logins, we need to let the exceptions bubble up.

File: src/Cms/Auth.php [hunk 4]
```diff
@@ -158,24 +157,34 @@ public function impersonate(string $who = null)
      */
     public function ipHash(): string
     {
-        return hash('sha256', $this->kirby->visitor()->ip());
+        $hash = hash('sha256', $this->kirby->visitor()->ip());
+
+        // only use the first 50 chars to ensure privacy
+        return substr($hash, 0, 50);
```
Comment (bastianallgeier) [inline src/Cms/Auth.php:161]: Good one!

**Review by bastianallgeier — CHANGES_REQUESTED** [review]
I think we should add proper tests for this entire PR to avoid further problems with the login protection.

File: config/api/routes/auth.php (hunk 1 above)
Comment (lukasbestle) [reply config/api/routes/auth.php]: No, that feature is still there (see the three lines above). I have only cleaned up the code a bit – it still works exactly the same way.

File: i18n/translations/en.json (hunk 2 above)
Comment (lukasbestle) [reply i18n/translations/en.json:172]: Before I removed it, I searched for `user.undefined` across the codebase and didn't find any remaining match. The error message is a duplicate of `error.user.notFound`.

File: src/Cms/Auth.php (hunk 3 above)
Comment (lukasbestle) [reply src/Cms/Auth.php]: That one was like this before as well. But I will see if I can improve on this further.

Comment (lukasbestle) [reply src/Cms/Auth.php]: Fixed now (not yet pushed).

**Comment by lukasbestle** [comment]
@bastianallgeier I have implemented all your feedback besides the removed translation string. Feel free to revert that change in case we still use it somewhere and I missed it or in case we will have a use for that variant without the email address placeholder later.

**Review by bastianallgeier — APPROVED** [review]
Looks great! Thanks for all the tests.

**Pull request merged**